### PR TITLE
mon,cephfs: require confirmation when changing max_mds on unhealthy cluster

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Fixes: https://tracker.ceph.com/issues/66301

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_admin.py [hunk 1]
```diff
@@ -2659,3 +2659,28 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
+    def test_fs_set_max_mds_when_unhealthy_with_confirm(self):
+        '''
+        Test that command "ceph fs set max_mds <num> --yes-i-really-mean-it"
+        run successfully when cluster is unhealthy.
+        '''
+        self.gen_health_warn_mds_cache_oversized()
+
+        self.fs.set_max_mds(2, confirm=True)
+        self.fs.getinfo()
+        self.assertEqual(self.fs.get_var('max_mds'), 2)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_admin.py:2686]: Move the qa tests to a separate commit?

File: qa/tasks/cephfs/test_admin.py [hunk 2]
```diff
@@ -2659,3 +2659,28 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
+        self.assertEqual(cfe.exception.exitstatus, errno.EPERM)
+
+    def test_fs_set_max_mds_when_unhealthy_with_confirm(self):
+        '''
+        Test that command "ceph fs set max_mds <num> --yes-i-really-mean-it"
+        run successfully when cluster is unhealthy.
+        '''
+        self.gen_health_warn_mds_cache_oversized()
+
+        self.fs.set_max_mds(2, confirm=True)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_admin.py:2684]: You could also write a test case using `confirm=True` when the cluster is healthy and it passes.

**Comment by joscollin** [comment]
@rishabh-d-dave  Otherwise, LGTM

File: qa/tasks/cephfs/test_admin.py [hunk 3]
```diff
@@ -2659,3 +2659,28 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
         self.run_ceph_cmd(f'mds fail {mds1_id} --yes-i-really-mean-it')
         self.run_ceph_cmd(f'mds fail {mds2_id} --yes-i-really-mean-it')
+
+
+class TestFSSet(TestAdminCommands):
+
+    def test_fs_set_max_mds_when_unhealthy_without_confirm(self):
+        '''
+        Test that command "ceph fs set max_mds <num>" without the confirmation
+        flag (--yes-i-really-mean-it) fails when cluster is unhealthy.
```
Comment (joscollin) [inline qa/tasks/cephfs/test_admin.py:2669]: It is `ceph fs set <fs_name> max_mds <num>`

File: src/mon/FSCommands.cc [hunk 4]
```diff
@@ -385,6 +384,16 @@ class SetHandler : public FileSystemCommandHandler
...
+  bool confirm = false;
+  cmd_getval(cmdmap, "yes_i_really_mean_it", confirm);
+  if (var == "max_mds" && !confirm && mon->mdsmon()->has_health_warnings()) {
+    ss << "FS has health warnings and modifying value of FS setting variable "
+       << "'max_mds' will not help troubleshoot or recover from one or more "
+       << "health warnings. If you wish to proceed, run again with "
+       << "--yes-i-really-mean-it";
+    return -EPERM;
+  }
+
```
Comment (joscollin) [inline src/mon/FSCommands.cc:396]: Needs doc update for `--yes-i-really-mean-it`

**Review by joscollin — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/filesystem.py [hunk 5]
```diff
@@ -640,8 +640,11 @@ def set_down(self, down=True):
     def set_joinable(self, joinable=True):
         self.set_var("joinable", joinable)
 
-    def set_max_mds(self, max_mds):
-        self.set_var("max_mds", "%d" % max_mds)
+    def set_max_mds(self, max_mds, confirm=False):
+        if not confirm:
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py:644]: reverse if, always prefer avoid using `not X` if an `else:` is present.

File: qa/tasks/cephfs/test_admin.py [hunk 6]
```diff
@@ -2659,3 +2659,28 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
+        '''
+        self.gen_health_warn_mds_cache_oversized()
+
+        with self.assertRaises(CommandFailedError) as cfe:
+            self.fs.set_max_mds(2)
+        self.assertEqual(cfe.exception.exitstatus, errno.EPERM)
```
Comment (batrick) [inline qa/tasks/cephfs/test_admin.py:2675]: test trim too as that's the more important one.

File: qa/tasks/cephfs/filesystem.py (hunk 5 above)
Comment (batrick) [reply qa/tasks/cephfs/filesystem.py:644]: Also, this change may break numerous tests. I wonder if we should, for our testing, make `confirm=True` the default.

File: src/mon/FSCommands.cc [hunk 7]
```diff
@@ -385,6 +384,16 @@ class SetHandler : public FileSystemCommandHandler
...
 
+  bool confirm = false;
+  cmd_getval(cmdmap, "yes_i_really_mean_it", confirm);
+  if (var == "max_mds" && !confirm && mon->mdsmon()->has_health_warnings()) {
+    ss << "FS has health warnings and modifying value of FS setting variable "
+       << "'max_mds' will not help troubleshoot or recover from one or more "
+       << "health warnings. If you wish to proceed, run again with "
+       << "--yes-i-really-mean-it";
+    return -EPERM;
+  }
```
Comment (batrick) [inline src/mon/FSCommands.cc:395]:
```suggestion
  if (var == "max_mds" && !confirm && mon->mdsmon()->has_health_warnings()) {
    ss << "One or more file system health warnings are present. Modifying the file system setting variable "
              "\"max_mds\" will not help troubleshoot or recover from these warnings and may further destabilize the system. "
              "If you really wish to proceed, run again with --yes-i-really-mean-it";
    return -EPERM;
  }
```

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/filesystem.py (hunk 5 above)
Comment (joscollin) [reply qa/tasks/cephfs/filesystem.py:644]: No, it doesn't seem to break any existing tests, as `self.fs.set_max_mds(<num>)` is still supported.

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/filesystem.py]: Added both the changes.

File: qa/tasks/cephfs/test_admin.py (hunk 6 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:2675]: Done.

File: src/mon/FSCommands.cc (hunk 7 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc]: Done.

File: qa/tasks/cephfs/test_admin.py (hunk 3 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:2669]: True, fixed.

File: src/mon/FSCommands.cc (hunk 4 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc:397]: Done.

File: qa/tasks/cephfs/test_admin.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:2684]: That's a good idea. I've added it.

File: qa/tasks/cephfs/test_admin.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:2686]: Yeah, forgot to do so. Done now. Thanks!

File: src/mon/FSCommands.cc (hunk 7 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc]: I've `s/may further destabilize the system/may further destabilize the cluster`. Let me know what you think.

**Comment by rishabh-d-dave** [comment]
@joscollin @batrick Thanks for the quick review. I've made the recommended changes now. PTAL.

File: src/mon/MDSMonitor.cc [hunk 8]
```diff
@@ -1540,7 +1540,7 @@ bool MDSMonitor::prepare_command(MonOpRequestRef op)
   }
 }
 
-bool MDSMonitor::has_health_warnings(vector<mds_metric_t> warnings)
+bool MDSMonitor::has_one_of_these_health_warnings(vector<mds_metric_t> warnings)
```
Comment (vshankar) [inline src/mon/MDSMonitor.cc:1543]: nit: `has_health_warnings` seems perfectly accurate to me.

File: src/mon/MDSMonitor.cc [hunk 9]
```diff
@@ -1557,6 +1557,16 @@ bool MDSMonitor::has_one_of_these_health_warnings(vector<mds_metric_t> warnings)
   return false;
 }
 
+bool MDSMonitor::has_health_warnings()
```
Comment (vshankar) [inline src/mon/MDSMonitor.cc:1560]: nit: `has_any_health_warning()` and keep `has_health_warnings` as it is.

File: src/mon/FSCommands.cc [hunk 10]
```diff
@@ -384,6 +384,17 @@ class SetHandler : public FileSystemCommandHandler
       return -EINVAL;
     }
 
+  bool confirm = false;
+  cmd_getval(cmdmap, "yes_i_really_mean_it", confirm);
+  if (var == "max_mds" && !confirm && mon->mdsmon()->has_health_warnings()) {
+    ss << "One or more file system health warnings are present. Modifying "
```
Comment (vshankar) [inline src/mon/FSCommands.cc:390]: nit: One or more file systems have health warnings. Modifying "max_mds" may destabilize the system further. Use `--yes...` to override.

File: qa/tasks/cephfs/test_admin.py [hunk 11]
```diff
@@ -2659,3 +2659,65 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
+        self.fs.getinfo()
+        self.assertEqual(self.fs.get_var('max_mds'), 2)
+
+    def test_when_healthy_with_confirm(self):
+        '''
+        Test that command "ceph fs set <fsname> max_mds <num>
+        --yes-i-really-mean-it" runs successfully also when cluster is
+        healthy.
+        '''
+        self.gen_health_warn_mds_trim()
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:2719]: err... Why is this generating a health warning when the test is supposed to validate the command without any health warning.

File: doc/cephfs/troubleshooting.rst [hunk 12]
```diff
@@ -128,6 +128,11 @@ things to do:
 
   That prevents any clients from establishing new sessions with the MDS.
 
+* **Dont tweak max_mds** Changing value of ``max_mds`` FS setting variable is
```
Comment (vshankar) [inline doc/cephfs/troubleshooting.rst:131]: I would add this as a precautionary note rather than an assertive statement.

File: qa/tasks/cephfs/test_admin.py [hunk 13]
```diff
@@ -2659,3 +2659,65 @@ def test_with_health_warn_with_2_active_MDSs(self):
...
+
+    def test_when_unhealthy_with_confirm(self):
+        '''
+        Test that command "ceph fs set <fsname> max_mds <num>
+        --yes-i-really-mean-it" runs successfully when cluster is unhealthy.
+        '''
+        self.gen_health_warn_mds_cache_oversized()
+
+        self.fs.set_max_mds(2, confirm=True)
+        self.fs.getinfo()
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:2686]: Why is this required?

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/mon/MDSMonitor.cc [hunk 14]
```diff
@@ -1557,6 +1557,16 @@ bool MDSMonitor::has_health_warnings(vector<mds_metric_t> warnings)
...
+{
+  for (auto& [gid, health] : pending_daemon_health) {
```
Comment (vshankar) [inline src/mon/MDSMonitor.cc:1562]: nit: use `std::any_of`.

https://en.cppreference.com/w/cpp/algorithm/all_any_none_of

File: src/mon/FSCommands.cc (hunk 10 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc]: I had something similar, current version was recommended by Patrick.

File: qa/tasks/cephfs/test_admin.py (hunk 11 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py]: Right, removed.

File: src/mon/MDSMonitor.cc (hunk 9 above)
Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc:1560]: I originally thought the same the but I perfer current version as I find it clearer.

File: doc/cephfs/troubleshooting.rst (hunk 12 above)
Comment (rishabh-d-dave) [reply doc/cephfs/troubleshooting.rst:131]: Please suggest better a wording. `Careful when tweakig max_mds`?

**Comment by rishabh-d-dave** [comment]
.

File: qa/tasks/cephfs/test_admin.py (hunk 13 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:2686]: Not really required. Removed.

File: PendingReleaseNotes [hunk 15]
```diff
@@ -12,6 +12,11 @@
   of the column showing the state of a group snapshot in the unformatted CLI
   output is changed from 'STATUS' to 'STATE'. The state of a group snapshot
   that was shown as 'ok' is now shown as 'complete', which is more descriptive.
+* CephFS: Modifying FS setting variable "max_mds" when a cluster is unhealthy
+  now requires users to pass the confirmation flag (--yes-i-really-mean-it).
+  This has been added as a precaution to tell user that modifying "max_mds"
+  value won't help with troublehshooting or recovery effort and instead might
+  destabilize the cluster further.
```
Comment (anthonyeleven) [inline PendingReleaseNotes:19]: Modifying the ``max_mds`` setting
confirmation flag ``--yes-i-really-mean-it``
as a precaution that modifying ``max_mds`` may not improve troubleshooting or recovery efforts and might additionally destabilize the cluster.

File: doc/cephfs/administration.rst [hunk 16]
```diff
@@ -61,10 +61,16 @@ is a subset of the same information from the ``ceph fs dump`` command.
...
 
 Change a setting on a file system. These settings are specific to the named
-file system and do not affect other file systems.
+file system and do not affect other file systems. Confirmation flag is only
+needed for changing ``max_mds`` when cluster is unhealthy.
+
+.. note:: Confirmation flag for changing ``max_mds`` FS setting variable when
+   cluster is unhealthy has been added as a precaution to tell user that
+   changing ``max_mds`` during troubleshooting or recovery won't help and
+   might worsen the system state.
```
Comment (anthonyeleven) [inline doc/cephfs/administration.rst:73]: See above

File: doc/cephfs/troubleshooting.rst (hunk 12 above)
Comment (anthonyeleven) [reply doc/cephfs/troubleshooting.rst:131]: See wording above

File: src/mon/FSCommands.cc [hunk 17]
```diff
@@ -385,6 +384,17 @@ class SetHandler : public FileSystemCommandHandler
       return -EINVAL;
     }
 
+  bool confirm = false;
+  cmd_getval(cmdmap, "yes_i_really_mean_it", confirm);
+  if (var == "max_mds" && !confirm && mon->mdsmon()->has_health_warnings()) {
+    ss << "One or more file systems health warnings are present. Modifying "
```
Comment (anthonyeleven) [inline src/mon/FSCommands.cc:390]: system

File: src/mon/FSCommands.cc [hunk 18]
```diff
@@ -385,6 +384,17 @@ class SetHandler : public FileSystemCommandHandler
...
+       << "the file system setting variable \"max_mds\" will not help "
```
Comment (anthonyeleven) [inline src/mon/FSCommands.cc:391]: s/will/may/

**Review by anthonyeleven — COMMENTED** [review]
suggestions

File: src/mon/FSCommands.cc (hunk 17 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc:390]: It should be `file systems` IMO since "more" comes right before.

Comment (anthonyeleven) [reply src/mon/FSCommands.cc:390]: "More" as I read it modifies "warnings", and "file system" is also used as an adjective to modify "warnings".

File: PendingReleaseNotes (hunk 15 above)
Comment (rishabh-d-dave) [reply PendingReleaseNotes]: Done.

File: src/mon/FSCommands.cc (hunk 17 above)
Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc:390]: You are right.

Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc:390]: I was reading it only partially due to a last review that asked me to `s/system/systems`. I should've read it full statement. Thanks!

Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc]: Fixed.

Comment (rishabh-d-dave) [reply src/mon/FSCommands.cc]: @anthonyeleven BTW, it would be great if you would  use
```suggestion
    ss1 << "One or more file systems health warnings are present. Modifying "
```

Comment (anthonyeleven) [reply src/mon/FSCommands.cc]: If you like.  I've found that accepting those via the web interface complicates things, ymmv.

**Comment by rishabh-d-dave** [comment]
@vshankar i've made the changes last week that were recommended by you. PTAL when you find some time.  :)

**Comment by vshankar** [comment]
> @vshankar i've made the changes last week that were recommended by you. PTAL when you find some time.  :)

I'll have a look tomorrow. Meanwhile requesting a review from @ceph/cephfs.

**Comment by rishabh-d-dave** [comment]
> > @vshankar i've made the changes last week that were recommended by you. PTAL when you find some time.  :)
> 
> I'll have a look tomorrow. Meanwhile requesting a review from @ceph/cephfs.

Thank you!

File: src/mon/MDSMonitor.cc (hunk 8 above)
Comment (vshankar) [reply src/mon/MDSMonitor.cc:1543]: @rishabh-d-dave - `has_health_warnings()` seems perfectly fine to me, so why change it?

File: src/mon/MDSMonitor.cc (hunk 9 above)
Comment (vshankar) [reply src/mon/MDSMonitor.cc:1560]: An earlier commit changes has_health_warnings to has_one_of_these_health_warnings and this commit introduces has_health_warnings. I don't understand why. Suggest change this to has_any_health_warning and not change the original has_health_warnings.

**Review by vshankar — COMMENTED** [review]
Minor nit, otherwise LGTM.

File: src/mon/MDSMonitor.cc (hunk 9 above)
Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc:1560]: Unlike `has_health_warnings()`, `has_one_of_these_health_warnings()` tells that this method looks for specific health warnings that have will be passed by the caller. I like this emphasis since this is more precise and makes code succinct and thereby easier to read.

Comment (vshankar) [reply src/mon/MDSMonitor.cc:1560]: This is fine if its pythonic code where the arg tells nothing about its type -- even that's changing now with `Typing`. Unless the function name does not make any sense *at all*, I'd appreciate it to be left alone.

Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc:1560]: okay.

**Comment by rishabh-d-dave** [comment]
Had a discussion wit Venky here, have made the relevant changes.

File: src/mon/MDSMonitor.cc (hunk 9 above)
Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc]: Made that change. Marking this as resolved.

File: src/mon/MDSMonitor.cc [hunk 19]
```diff
@@ -1557,6 +1557,13 @@ bool MDSMonitor::has_health_warnings(vector<mds_metric_t> warnings)
   return false;
 }
 
+bool MDSMonitor::has_any_health_warning()
+{
+  return std::any_of(
+    pending_daemon_health.begin(), pending_daemon_health.end(),
+    [](auto& it) { return it.second.metrics.size() > 0 ? true : false; });
```
Comment (vshankar) [inline src/mon/MDSMonitor.cc:1564]: OK. This is new (I think). Wasn't this just `!it.second.metrics.empty()`?

**Review by vshankar — APPROVED** [review]
Otherwise LGTM.

File: src/mon/MDSMonitor.cc (hunk 19 above)
Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc:1564]: Nope. It's exactly same as before.

Comment (vshankar) [reply src/mon/MDSMonitor.cc:1564]: kk, then that was a different change then. Anyway, please use `!it.second.metrics.empty()` and run this through fs suite 👍

Comment (rishabh-d-dave) [reply src/mon/MDSMonitor.cc]: Okay. Done.

**Comment by rishabh-d-dave** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68354](https://tracker.ceph.com/issues/68354).

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Review by rishabh-d-dave — COMMENTED** [review]
QA run was successful - https://tracker.ceph.com/projects/cephfs/wiki/Main#2024-Oct-18

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
Ready for merge but waiting for Ceph API CI job to pass
